scylla/src/transport/load_balancing/default.rs

```diff
@@ -123,7 +134,26 @@ or refrain from preferring datacenters (which may ban all other datacenters, if
 
         // If token is available, get a shuffled list of alive replicas.
         let maybe_replicas = if let Some(ts) = &routing_info.token_with_strategy {
-            let local_replicas = self.shuffled_replicas(ts, true, Self::is_alive, cluster);
+            let local_rack_replicas = self.replicas(
+                ts,
+                true,
+                |node| {
+                    Self::is_alive(node)
+                        && self.preferred_rack.is_some()
+                        && DefaultPolicy::rack_predicate(node, &self.preferred_rack)
```

Comment:
I would prefer to include checking if `preferred_rack` is not `None` to `rack_predicate`. It would reduce the complexity of this expression.